cdap-app-fabric/src/main/java/co/cask/cdap/extension/ExtensionLoader.java

```diff
@@ -0,0 +1,227 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.extension;
+
+import co.cask.cdap.common.utils.DirUtils;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Function;
+import com.google.common.base.Objects;
+import com.google.common.base.Splitter;
+import com.google.common.base.Throwables;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.ServiceLoader;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * A class to maintain a cache of extensions available in a configured extensions directory. It uses the Java
+ * {@link ServiceLoader} architecture to load extensions from the specified directory. It first tries to load
+ * extensions using a {@link URLClassLoader} consisting of all jars in the configured extensions directory. For unit
+ * tests, it loads extensions using the system classloader.
+ *
+ * The supported directory structure is:
+ * <pre>
+ *   [extensions-directory]/[extensions-module-directory]/
+ *   [extensions-directory]/[extensions-module-directory]/file-containing-extensions-1.jar
+ *   [extensions-directory]/[extensions-module-directory]/file-containing-extensions-2.jar
+ *   ...
+ *   [extensions-directory]/[extensions-module-directory]/file-containing-extensions-n.jar
+ * </pre>
+ *
+ * Each extensions jar file above can contain multiple extensions.
+ * 
+ * @param <CACHE_KEY> the key type stored in the cache
+ * @param <CACHE_VALUE> the value type stored in the cache
+ */
+public class ExtensionLoader<CACHE_KEY, CACHE_VALUE> {
+  private static final Logger LOG = LoggerFactory.getLogger(ExtensionLoader.class);
+
+  private final List<String> extDirs;
+  private final Class<CACHE_VALUE> cacheValueClass;
+  // The ServiceLoader that loads extension implementation from the CDAP system classloader.
+  private final ServiceLoader<CACHE_VALUE> systemExtensionLoader;
+  private final Function<CACHE_VALUE, Set<CACHE_KEY>> extensionToKeys;
+  private final LoadingCache<CACHE_KEY, CACHE_VALUE> extensionsCache;
+  private final CACHE_VALUE defaultExtension;
+
+  public ExtensionLoader(String extDirs, Function<CACHE_VALUE, Set<CACHE_KEY>> extensionToKeys,
```

Comment:
Its because this class uses cache terminology which like you said it should not. This is basically a ``Function`` to map an extension, to its supported types. e.g. ``SparkRuntimeProvider`` to ``ProgramType.SPARK``. I'll update the name, and include a javadoc.